### DISCUSSION TITLE
Can your code also use the MOTOR PWM of ESP32

### DISCUSSION THREAD

**Discussion by evup4s** [Dlloydev/ESP32-ESP32S2-AnalogWrite#21] [Q&A]
If i am correct this code is (only) using the  'LEDC' PWM controller of the ESP32. I was wondering if (in the future?) it could also use the MOTOR PWM modulator of the ESP32, because i need to (accurately) control the frame speed of a (timing) camera, and for (exact) synchronisation of the camera i need more options than the freq, phase and duty cycle of an output signal of the ESP32. In fact, i need to have an interrupt generated at every timer period, and occasionally change the timer period on the spot, from this interrupt routine, or from somewhere else, with semaphores.
I managed to do this with an AT-MEGA2560 board, simply because i found the "Timer Output COMPARE" registers of that Atmel processor, which not only define the timer value to generate a (pos,neg,toggle) edge on an output pin but also generate an interrupt. Wish this could be done with an ESP32, because the T-Beam product of LilyGo/TTGo is exactly what i need for my project, including Wifi, BT and LoRa comm, Oled, battery, Neo gps, USB, etc...
Or do you know any other library for the MOTOR PWM control of the ESP32?

**Comment by Dlloydev** [comment]
I won't be adding MCPWM control to this library, but it seems like you might not actually need to use this hardware feature.
There isn't interrupt methods used in this library yet, but I'm open to pull requests.
Perhaps something in the examples [here](https://www.visualmicro.com/page/Timer-Interrupts-Explained.aspx) could help with your application.

**Comment by Dlloydev** [comment]
I've been working on a major update to the library (version 4.0.0) which will have some significant new capabilities using the LEDC PWM controller of the ESP32/S2/S3/C3. For the ESP32, all 16 PWM channels will be available.

The new version will exploit the capabilities of the LEDC controller to add PWM synchronization. Its not perfect, but multiple PWM signals can be synchronized to within 1μs. I think this pushes the LEDC capabilities much closer to that of the MCPWM controller and it has the capacity to synchronize more signals than MCPWM.

#### ESP32 PWM Synchronization:

- New capability to pause/resume the timers to allow PWM startup synchronization.
- Up to 8 PWM signal pairs are inherently synchronized. Channels 0-1, 2-3, 4-5, 6-7, 8-9, 10-11, 12-13, 14-15. 
- Each channel pair uses a different timer and/or speed mode. I've found that there's a consistent delay of about 11μs between each signal pair.  This can be essentially calibrated out by using the hpoint (phase) value to delay the PWM signal start. The result is the alignment of each signal pair (timer to timer synchronization).

Here's the logic trace of 8 PWM signals (synchronized) at 25kHz frequency and 10-bit resolution:

![image](https://user-images.githubusercontent.com/63488701/203777997-e0daa4e7-0252-42f1-9262-8c25e37b0588.png)

Zooming in, the timer to timer synchronization (after phase delay calibration) was only100ns:

![image](https://user-images.githubusercontent.com/63488701/203779521-e22eb2fb-4a34-466d-85cf-625c1a5af890.png)

Zooming in to max, here we can see perfect synchronization between a signal pair because they use the same timer:

![image](https://user-images.githubusercontent.com/63488701/203780148-32ca0871-99d2-49af-8b9f-a0d302858183.png)

I should have the new release ready within the next week ... still need to update documentation.
